fireworks/core/mongo_launchpad.py

```diff
@@ -0,0 +1,1407 @@
+
+    def to_dict(self):
+        """
+        Note: usernames/passwords are exported as unencrypted Strings!
+        """
+        return {
+            'host': self.host,
+            'port': self.port,
+            'name': self.name,
+            'username': self.username,
+            'password': self.password,
+            'logdir': self.logdir,
+            'strm_lvl': self.strm_lvl,
+            'user_indices': self.user_indices,
+            'wf_user_indices': self.wf_user_indices,
+            'ssl': self.ssl,
+            'ssl_ca_certs': self.ssl_ca_certs,
+            'ssl_certfile': self.ssl_certfile,
+            'ssl_keyfile': self.ssl_keyfile,
+            'ssl_pem_passphrase': self.ssl_pem_passphrase,
+            'authsource': self.authsource}
+
+    def update_spec(self, fw_ids, spec_document, mongo=False):
+        """
+        Update fireworks with a spec. Sometimes you need to modify a firework in progress.
+
+        Args:
+            fw_ids [int]: All fw_ids to modify.
+            spec_document (dict): The spec document. Note that only modifications to
+                the spec key are allowed. So if you supply {"_tasks.1.parameter": "hello"},
+                you are effectively modifying spec._tasks.1.parameter in the actual fireworks
+                collection.
+            mongo (bool): spec_document uses mongo syntax to directly update the spec
+        """
+        if mongo:
+            mod_spec = spec_document
+        else:
+            mod_spec = {"$set": {("spec." + k): v for k, v in spec_document.items()} }
+
+        allowed_states = ["READY", "WAITING", "FIZZLED", "DEFUSED", "PAUSED"]
+        self.fireworks.update_many({'fw_id': {"$in": fw_ids},
+                                    'state': {"$in": allowed_states}}, mod_spec)
+        for fw in self.fireworks.find({'fw_id': {"$in": fw_ids}, 'state': {"$nin": allowed_states}},
+                                      {"fw_id": 1, "state": 1}):
+            self.m_logger.warning("Cannot update spec of fw_id: {} with state: {}. "
+                               "Try rerunning first".format(fw['fw_id'], fw['state']))
+
+    @classmethod
+    def from_dict(cls, d):
+        port = d.get('port', None)
+        name = d.get('name', None)
+        username = d.get('username', None)
+        password = d.get('password', None)
+        logdir = d.get('logdir', None)
+        strm_lvl = d.get('strm_lvl', None)
+        user_indices = d.get('user_indices', [])
+        wf_user_indices = d.get('wf_user_indices', [])
+        ssl = d.get('ssl', False)
+        ssl_ca_certs = d.get('ssl_ca_certs', d.get('ssl_ca_file', None))  # ssl_ca_file was the old notation for FWS < 1.5.5
+        ssl_certfile = d.get('ssl_certfile', None)
+        ssl_keyfile = d.get('ssl_keyfile', None)
+        ssl_pem_passphrase = d.get('ssl_pem_passphrase', None)
+        authsource= d.get('authsource', None)
+        return LaunchPad(d['host'], port, name, username, password,
```

Comment:
`return MongoLaunchpad(...`

Reply:
Fixed in 5ddbea4e8f4696b2c20bf453397932146d7a6536